cmd/nerdctl/run_linux.go

```diff
@@ -217,3 +210,90 @@ func withOOMScoreAdj(score int) oci.SpecOpts {
 		return nil
 	}
 }
+
+func generatePIDOpts(cmd *cobra.Command) (oci.SpecOpts, error) {
+	opts := make([]oci.SpecOpts, 0)
+
+	pid, err := cmd.Flags().GetString("pid")
+	if err != nil {
+		return nil, err
+	}
+	pid = strings.ToLower(pid)
+
+	switch pid {
+	case "":
+		// do nothing
+	case "host":
+		opts = append(opts, oci.WithHostNamespace(specs.PIDNamespace))
+		if rootlessutil.IsRootless() {
+			opts = append(opts, withBindMountHostProcfs)
+		}
+	default: // container:<id|name>
+		client, ctx, cancel, err := newClient(cmd)
+		if err != nil {
+			return nil, err
+		}
+
+		defer cancel()
+
+		parsed := strings.Split(pid, ":")
+		if len(parsed) < 2 || parsed[0] != "container" {
+			return nil, fmt.Errorf("invalid pid namespace. Set --pid=[host|container:<name|id>])")
+		}
+
+		walker := &containerwalker.ContainerWalker{
+			Client: client,
+			OnFound: func(ctx context.Context, found containerwalker.Found) error {
+				if found.MatchCount > 1 {
+					return fmt.Errorf("multiple IDs found with provided prefix: %s", found.Req)
+				}
+
+				nc, err := containerinspector.Inspect(ctx, found.Container)
+				if err != nil {
+					return err
+				}
+
+				if nc.Process.Status.Status != containerd.Running {
+					return fmt.Errorf("shared container is not running")
+				}
+
+				spec, err := found.Container.Spec(ctx)
+				if err != nil {
+					return err
+				}
+
+				isHost := true
+				for _, n := range spec.Linux.Namespaces {
+					if n.Type == specs.PIDNamespace {
+						isHost = false
+					}
+				}
+				if isHost {
+					opts = append(opts, oci.WithHostNamespace(specs.PIDNamespace))
+					if rootlessutil.IsRootless() {
+						opts = append(opts, withBindMountHostProcfs)
+					}
+				} else {
+					ns := specs.LinuxNamespace{
+						Type: specs.PIDNamespace,
+						Path: fmt.Sprintf("/proc/%d/ns/pid", nc.Process.Pid),
+					}
+					opts = append(opts, oci.WithLinuxNamespace(ns))
+				}
+
+				return nil
+			},
+		}
+
+		containerName := parsed[1]
+		matchedCount, err := walker.Walk(ctx, containerName)
+		if err != nil {
+			return nil, err
+		}
+		if matchedCount < 1 {
+			return nil, fmt.Errorf("no such container: %s", containerName)
+		}
+	}
+
+	return oci.Compose(opts...), nil
```

Comment:
why compose?

Reply:
no need. I'll fix the function returns []oci.Specs.